daemons/controld/controld_te_events.c

```diff
@@ -20,6 +20,57 @@
 #include <crm/common/attrd_internal.h>
 #include <crm/common/ipc_attrd_internal.h>
 
+/*!
+ * \internal
+ * \brief Action numbers of outside events processed in current update diff
+ *
+ * This table is to be used as a set. It should be empty when the transitioner
+ * begins processing a CIB update diff. It ensures that if there are multiple
+ * events (for example, "_last_0" and "_last_failure_0") for the same action,
+ * only one of them updates the failcount. Events that originate outside the
+ * cluster can't be confirmed, since they're not in the transition graph.
+ */
+static GHashTable *outside_events = NULL;
+
+/*!
+ * \internal
+ * \brief Destroy the hash table containing action numbers of outside events
+ */
+void
+controld_destroy_outside_event_table(void)
+{
+    if (outside_events != NULL) {
+        g_hash_table_destroy(outside_events);
+        outside_events = NULL;
+    }
+}
+
+/*!
+ * \internal
+ * \brief Add an outside event's action number to a set
+ *
+ * \return Standard Pacemaker return code. Specifically, \p pcmk_rc_ok if the
+ *         event was not already in the set, or \p pcmk_rc_already otherwise.
+ */
+static int
+record_outside_event(int action_num)
+{
+    int *action_num_copy = NULL;
+
+    if (outside_events == NULL) {
+        outside_events = g_hash_table_new_full(g_int_hash, g_int_equal, free,
+                                               free);
+    }
+    action_num_copy = calloc(1, sizeof(int));
+    CRM_ASSERT(action_num_copy != NULL);
+    *action_num_copy = action_num;
+
+    if (g_hash_table_add(outside_events, action_num_copy)) {
```

Comment:
You don't want free,free if you're using the same memory for key and value. That might be related to your issue.

FYI g_hash_table_add() and similar returned void before 2.40, so we lucked out being able to use this :)

Reply:
> That might be related to your issue.

Probably, thank you I'll try that. It didn't click that this was the same memory, but that makes sense since it's all pointer-based.

> FYI g_hash_table_add() and similar returned void before 2.40, so we lucked out being able to use this :)

Yep, I saw the version-added note :) Really glad we bumped our GLib dependency a while back.